Auth\Basic: add input validation to constructor

jrfnl opened the pull request:
The `Basic::__construct()` method expects to receive an array of exactly two element and did validate the input received, but only threw an exception when the passed parameter did not match the expected element count, not when the passed parameter wasn't an array.

At the same time, the default value of `$args` is `null`, not an empty array and the `$user` and `$pass` properties are both public.

The only reason I can think of to have these properties `public` and the default value set to `null`, is to allow for delayed setting of the `$user` and `$pass`, after the class has already been instantiated.

To maintain the existing behaviour, but still add input validation, an exception is now thrown when the `$args` parameter is neither an array, nor `null`.

Includes;
* A dedicated test for the input validation.
* An additional test with Basic authentication with an instantiated class, but with delayed setting of the `$user` and `$pass` directly on the instantiated object. This test is solely added to document that this is a supported feature (and to safeguard against regressions in that respect).

jrfnl committed:
Auth\Basic: add input validation to constructor

The `Basic::__construct()` method expects to receive an array of exactly two element and did validate the input received, but only threw an exception when the passed parameter did not match the expected element count, not when the passed parameter wasn't an array.

At the same time, the default value of `$args` is `null`, not an empty array and the `$user` and `$pass` properties are both public.

The only reason I can think of to have these properties `public` and the default value set to `null`, is to allow for delayed setting of the `$user` and `$pass`, after the class has already been instantiated.

To maintain the existing behaviour, but still add input validation, an exception is now thrown when the `$args` parameter is neither an array, nor `null`.

Includes;
* A dedicated test for the input validation.
* An additional test with Basic authentication with an instantiated class, but with delayed setting of the `$user` and `$pass` directly on the instantiated object. This test is solely added to document that this is a supported feature (and to safeguard against regressions in that respect).

schlessera commented on src/Auth/Basic.php:
```diff
@@ -39,8 +40,10 @@ class Basic implements Auth {
 	 * @param array|null $args Array of user and password. Must have exactly two elements
+	 *
+	 * @throws \WpOrg\Requests\Exception\InvalidArgument When the passed argument is not an array or null.
+	 * @throws \WpOrg\Requests\Exception                 On incorrect number of arguments (`authbasicbadargs`).
```
Should we immediately change this Exception into one that is more precise? I think it should be based on https://www.php.net/manual/en/class.badmethodcallexception.php. If we make sure that it extends `\WpOrg\Requests\Exception`, it shouldn't be a breaking change.

jrfnl replied:
Good idea. What about making it an `ArgumentCountException` though ? Kind of feels more appropriate ?
https://www.php.net/manual/en/class.argumentcounterror.php

schlessera replied:
That doesn't exist for PHP 5.6, though...

jrfnl replied:
> That doesn't exist for PHP 5.6, though.

True, but that shouldn't really matter if we extend the `\WpOrg\Requests\Exception` class anyhow.

> If we make sure that it extends `\WpOrg\Requests\Exception`, it shouldn't be a breaking change.

jrfnl replied:
I've added the `ArgumentCount` exception class + implemented use of it in this class + one other location in the code base where it seemed appropriate.

jrfnl committed:
Introduce new `ArgumentCount` Exception

... for use when an invalid number of arguments are received for a method.

Along the same lines as the `InvalidArgument` exception, this class contains a `create()` method to standardize the text of the exception message.

The exception extends the Request native base `Exception` to
* Allow for introducing this as a (more specific) replacement for some existing exceptions without BC-break.
* Allow for maintaining the "exception type" indicators as were used before.

Includes dedicated unit test.

jrfnl committed:
ArgumentCount: implement use of the new exception

This replaces two pre-existing generic Requests `Exception` usage instances with use of the new `ArgumentCount` exception.

Includes:
* Updating the parameter description for the `$args` parameter of the `Proxy\Http::__construct()` method as those docs were plain incorrect to begin with.
* Updated and improved documentation for these exceptions in the function doc blocks.
    Again, the documentation for the exception thrown in the `Proxy\Http` class was incorrect to begin with.
* Updated unit tests for the affected classes to reflect these changes.
